subiquity/tests/api/test_api.py

```diff
@@ -0,0 +1,521 @@
+            with async_timeout.timeout(_timeout):
+                return await coro(*args, **kwargs)
+        return run
+    return wrapper
+
+
+def json_print(json_data):
+    print(json.dumps(json_data, indent=4))
+
+
+def loads(data):
+    return json.loads(data) if data else None
+
+
+def dumps(data):
+    # if the data we're dumping is literally False, we want that to be 'false'
+    if data or type(data) is bool:
+        return json.dumps(data, separators=(',', ':'))
+    elif data is not None:
+        return '""'
+    else:
+        return data
+
+
+class ClientException(Exception):
+    pass
+
+
+class TestAPI(unittest.IsolatedAsyncioTestCase):
+    machine_config = 'examples/simple.json'
+    need_spawn_server = True
+
+    async def get(self, query, **kwargs):
+        return await self.request('GET', query, **kwargs)
+
+    async def post(self, query, data=None, **kwargs):
+        return await self.request('POST', query, data, **kwargs)
+
+    async def request(self, method, query, data=None, **kwargs):
+        params = {}
+        for key in kwargs:
+            params[key] = dumps(kwargs[key])
+        data = dumps(data)
+        info = f'{method} {query}'
+        if params:
+            for i, key in enumerate(params):
+                joiner = '?' if i == 0 else '&'
+                info += f'{joiner}{key}={params[key]}'
+        print(info)
+        async with self.session.request(method, f'http://a{query}',
+                                        data=data, params=params) as resp:
+            resp.raise_for_status()
+            content = await resp.content.read()
+            return loads(content.decode())
+
+    async def server_startup(self):
+        for _ in range(20):
+            try:
+                await self.get('/meta/status')
+                return
+            except aiohttp.client_exceptions.ClientConnectorError:
+                await asyncio.sleep(.5)
+        raise Exception('timeout on server startup')
+
+    async def server_shutdown(self, immediate=True):
+        try:
+            await self.post('/shutdown', mode='POWEROFF', immediate=immediate)
+            raise Exception('expected ServerDisconnectedError')
+        except aiohttp.client_exceptions.ServerDisconnectedError:
+            return
+
+    async def spawn_server(self):
+        env = os.environ.copy()
+        env['SUBIQUITY_REPLAY_TIMESCALE'] = '100'
+        cmd = 'python3 -m subiquity.cmd.server --dry-run --bootloader uefi' \
+              + ' --machine-config ' + self.machine_config
+        cmd = cmd.split(' ')
+        self.proc = await astart_command(cmd, env=env)
+        self.server = asyncio.create_task(self.proc.communicate())
+
+    async def asyncSetUp(self):
+        if self.need_spawn_server:
+            await self.spawn_server()
+
+        # setup client
+        conn = aiohttp.UnixConnector(path=socket_path)
+        self.session = aiohttp.ClientSession(connector=conn)
+        await self.server_startup()
```

Comment:
So I sort of hate using implicit setup for this. Would it be possible to have a helper method that each test can call to set up a server (that would of course arrange for the server to be torn down at the end of the test using addAsyncCleanup)?